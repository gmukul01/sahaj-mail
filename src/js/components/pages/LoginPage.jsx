import React, { useState, useEffect } from 'react';
import { Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';
import ReactLoading from 'react-loading';

import { useFormInput } from 'effects/useFormInput';

export const Loginpage = props => {
	const { setValue: setEmail, ...email } = useFormInput(''),
		{ setValue: setPassword, ...password } = useFormInput(''),
		[errorMessage, setErrorMessage] = useState(props.errorMessage);

	const { isUserLoggedIn, isLoading, fetchUserDetails, location } = props,
		{ from } = location.state || { from: undefined };

	useEffect(() => {
		setErrorMessage(props.errorMessage);
	}, [props.errorMessage]);

	if (isUserLoggedIn) {
		return <Redirect to={from || '/'} />;
	}

	const handleSubmit = e => {
		e.preventDefault();

		if (email.value === '' && password.value === '') setErrorMessage('Please enter Email & Password');
		else if (email.value === '') setErrorMessage('Please enter Email');
		else if (password.value === '') setErrorMessage('Please enter Password');
		else fetchUserDetails(email.value, password.value);
	};

	return (
		<section className="login">
			{isLoading ? (
				<ReactLoading type="spin" color="black" />
			) : (
				<form onSubmit={handleSubmit} className="login-card">
					<h1>Sign In</h1>
					<input type="text" {...email} placeholder="Username" />
					<input type="password" {...password} placeholder="Password" />
					<button type="submit">Login</button>
				</form>
			)}
			{errorMessage && <p className="error">{errorMessage}</p>}
		</section>
	);
};

Loginpage.propTypes = {
	isUserLoggedIn: PropTypes.bool,
	isLoading: PropTypes.bool,
	fetchUserDetails: PropTypes.func.isRequired,
	location: PropTypes.object.isRequired
};

export default React.memo(Loginpage);
