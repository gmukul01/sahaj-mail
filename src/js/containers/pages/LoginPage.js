import { connect } from 'react-redux';

import { fetchUserDetails } from 'actions/user';
import Loginpage from 'components/pages/Loginpage';

const mapStateToProps = ({ user: { name, isLoading, errorMessage } }) => ({
	isLoading,
	errorMessage,
	isUserLoggedIn: name ? true : false
});

export default connect(
	mapStateToProps,
	{
		fetchUserDetails
	}
)(Loginpage);
