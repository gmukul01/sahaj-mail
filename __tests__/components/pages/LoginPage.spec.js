import React from 'react';
import { render, fireEvent } from 'react-testing-library';
import 'react-testing-library/cleanup-after-each';

import LoginPage from 'components/pages/LoginPage';

describe('LoginPage Component', () => {
	const mockFetchUserDetails = jest.fn();
	const initialProps = {
		isLoading: false,
		fetchUserDetails: mockFetchUserDetails,
		location: {
			from: '/'
		}
	};

	it('should match snapshot', () => {
		const { container } = render(<LoginPage {...initialProps} />);
		expect(container).toMatchSnapshot();
	});

	it('should show error when all the fields are empty', () => {
		const { container, getByText } = render(<LoginPage {...initialProps} />);
		fireEvent.click(getByText('Login'));
		expect(container.querySelector('.error').textContent).toBe('Please enter Email & Password');
	});

	it('should show error when Password fields is empty', () => {
		const { container, getByPlaceholderText, getByText } = render(<LoginPage {...initialProps} />);
		fireEvent.change(getByPlaceholderText('Username'), { target: { value: 'DummyUserName' } });
		fireEvent.click(getByText('Login'));
		expect(container.querySelector('.error').textContent).toBe('Please enter Password');
	});

	it('should show error when Email fields is empty', () => {
		const { container, getByPlaceholderText, getByText } = render(<LoginPage {...initialProps} />);
		fireEvent.change(getByPlaceholderText('Password'), { target: { value: 'DummyPassword' } });
		fireEvent.click(getByText('Login'));
		expect(container.querySelector('.error').textContent).toBe('Please enter Email');
	});

	it('should call fetch Username use details if email and password is provided ', () => {
		const { getByPlaceholderText, getByText } = render(<LoginPage {...initialProps} />);
		fireEvent.change(getByPlaceholderText('Username'), { target: { value: 'DummyUserName' } });
		fireEvent.change(getByPlaceholderText('Password'), { target: { value: 'DummyPassword' } });
		fireEvent.click(getByText('Login'));
		expect(mockFetchUserDetails.mock.calls[0][0]).toBe('DummyUserName');
		expect(mockFetchUserDetails.mock.calls[0][1]).toBe('DummyPassword');
	});
});
