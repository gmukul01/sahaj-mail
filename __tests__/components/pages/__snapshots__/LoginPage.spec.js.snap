// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage Component should match snapshot 1`] = `
<div>
  <section
    class="login"
  >
    <form
      class="login-card"
    >
      <h1>
        Sign In
      </h1>
      <input
        placeholder="Username"
        type="text"
        value=""
      />
      <input
        placeholder="Password"
        type="password"
        value=""
      />
      <button
        type="submit"
      >
        Login
      </button>
    </form>
  </section>
</div>
`;

exports[`LoginPage Component should match snapshot when isUserLoggedIn 1`] = `<div />`;
